sklearn/tree/_criterion.pyx

```diff
@@ -454,7 +444,7 @@ cdef class ClassificationCriterion(Criterion):
 
                 for k in range(self.n_outputs):
                     label_index = (k * self.sum_stride +
-                                   <SIZE_t> y[i * self.y_stride + k])
+                                   <SIZE_t> self.y[i, k])
```

Comment:
Looking at the annotated html file we can see that using `self` generates some small Python interactions. I still don't understand why: this is a cdef class and `self.y` is properly typed, but I have observed this in #12807 as well.

Having only one Python interation is OK, but having interactions inside loops like here might be costly, so it's best to remove them. The workaround I found is to declare a local view and use it instead:

```python
def const DOUBLE_t [:, ::1] y = self.y
```

Sorry @adrinjalali I know I pretty much told you to use `self.` in https://github.com/scikit-learn/scikit-learn/pull/12886#issuecomment-450878358, but I wasn't aware of this at the time :s

Reply:
It's quite strange that the observation contradicts what we'd expect. I did change them and put the local variables in place, but the overhead of the local variable seems to be kinda much more than the little python interaction when using `self`. So I guess we should keep them as is, although along the way I did a bit more cleanups and the benchmarks are still intact.

Reply:
Okay,

I'll try to come up with a reproducing example and ask Cython folks why we're having Python interactions. Meanwhile I agree, let's keep it as is.

Thanks for double-checking!